### DISCUSSION TITLE
Bosch Home Connect Alt - confirm installation process

### DISCUSSION THREAD

**Discussion by llundberg57** [ekutner/home-connect-hass#83] [Community Support] [answered]
Hi,
I set up a developer account and went to the Register Application page.
It asks me to add:
Application ID - Is this something I can make up myself?

OAuth Flow - options are Authorization Code Grant Flow or Device Flow. Not sure what these mean. I assume that is important if I want to develop. What did you pick?

Redirect URL - Not sure what this is? I have a Home Connect account set up and working but I don’t know what the URL is? The way I access Bosch Home Connect is via the Android app.

I have started a thread on the Home Assistant forum for this but with limited responses - https://community.home-assistant.io/t/bosch-home-connect-alt-confirm-installation-process/436088

Apologies if these are dumb questions but the process is not clear to me and I cannot find any additional documentation.

Please help if you can, I have no where else to turn.

**Comment by ekutner** [comment]
Follow the instructions.

**Reply by llundberg57** [reply]
Is this really your response to someone asking for help?

I have managed to find answers to the questions above and as you probably implied, yes, it was my mistake.
Now I get;
error": "unauthorized_client",
"error_description": "Invalid client id"

I guess I will have to look elsewhere.

Thanks for your reply.

**Reply by Asdrubale1234** [reply]
any solutions? for invalid client id?

**Reply by mitch66692** [reply]
I got the same problem! Could you fix it?

**Comment by ekutner** [comment]
I think you are confusing me for the support team. I'm doing this on my spare time and the basic expectation when using open source software is to make the effort to look for existing information before asking for help. Your issue both the original and this one have been discussed in previous issues and also mentioned in the FAQ.

**Reply by llundberg57** [reply]
I understand that you are doing this in your spare time and not a support team. I do realize this.
I do apologize, I was frustrated because I wasn't getting anywhere. I also know that the previous issue was my fault and I found the solution eventually.
If the current issue is covered as well, I will do some more reading and hopefully find the answer. Sometimes it comes to a dead-end point and this is where I sometimes need just a pointer perhaps where to look. I am sure I am not the only one in that situation.
But each time I solve something, I get a little bit better.

Lastly, I also like to thank you for your efforts with this integration.

**Reply by llundberg57** [reply]
BTW, I respect your efforts and time and I don't expect you to do anything for nothing. I am willing to pay you for your assistance. If you tell me a figure, I am prepared to pay in advance via PayPal or whatever method is preferable to you. 

I am serious about this,  if this is OK with you,  please send me a private on llundberg57?@gmail.com (delete the? ).

**Reply by ekutner** [reply]
I appreciate the gesture but this is not about money, it's just about demonstrating you've made the effort to solve the problem yourself using the available information before asking for others to help.

Anyway, these problems are ALWAYS a result of not properly following the configuration procedure. The error message you are getting suggests you have the wrong client ID configured in your configuration.yaml, or that it doesn't match the secret key. Start by verifying that, then make sure you have properly configured the app on Home Connect as instructed in the instructions for the [built-in integration ](https://www.home-assistant.io/integrations/home_connect/). Also make sure that you have properly configured the URL for your HA installation in my.home-assistant.io

**Reply by llundberg57** [reply]
Thanks, 
I do understand and agree that one should always try and solve problems oneself in the first instance. What concerns me is what to do when those efforts lead to a dead-end. I'm very much a novice but I want and do try and learn.
But in your latest response I think you have given me some pointers to follow up on and this is kind of what I was hoping for, not the actual solution but a pointer where to look.

**Reply by llundberg57** [reply]
Hi again,
I apologise for this in advance as I know you must be getting sick of me. What I have done, I started by checking everything I have done so far and this is it:
Under pre-requisites for the basic Home-Connect:
1.	Visit https://developer.home-connect.com and sign up for a developer account - DONE
2.	Enter the email of your login for the original Home Connect App from Bosch/Siemens under “Default Home Connect User Account for Testing” in the sign up process - DONE
3.	Under Applications, register a new App:
•	Application ID: Home Assistant (or whatever name makes sense to you) - DONE (**Home Assistant**)
•	OAuth Flow: Authorization Code Grant Flow – DONE (added **Authorization Code Grant Flow**)
•	Redirect URL: https://my.home-assistant.io/redirect/oauth - DONE (added **https://my.home-assistant.io/redirect/oauth**)
One Time Token Mode is Disabled as the instructions made no mention of this.
Proof Key for Code Exchange is also Disabled as the instructions made no mention of this.


Then, in Home Assistant, I go Settings > Devices & Services > Add Integration > Home Connect > I then get a popup with ‘Open website’, I click on Open Website and I get a new web page with:
{
  "error": "unauthorized_client",
  "error_description": "Invalid client id",
  "state": "eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJmbG93X2lkIjoiODQwMjQxZmE1MDc4MGViMTIzYTE4YWE3MzJkZWI4NDkiLCJyZWRpcmVjdF91cmkiOiJodHRwczovL215LmhvbWUtYXNzaXN0YW50LmlvL3JlZGlyZWN0L29hdXRoIn0.W-R-yfHSVerqGZs4veCWb-nIaJad36vTWy2F1QPvNA8"
}

These are the first steps in the instructions for the basic Home Connect integration before doing any configuration and I have  already struck a problem. I believe I have followed the instructions right down to the letter.
I have no client ID in my configuration.yaml as the instructions did not say to add this. Nor have I added the secret key anywhere for the same reason.
For the my.home-assistant.io, I have **http://homeassistant.local:8123/**

If you have any pointers what I can look at, I would be so grateful.

**Comment by ekutner** [comment]
Well, you obviously need to set the client ID and secret in the configuration and those are clearly labeled as required parameters in the configuration section of the README for this integration.

**Reply by llundberg57** [reply]
Thank you for the prompt response.
I was just trying to follow the instructions for the basic integration to start with and there is no mention of this in there.
That being said, I have now added this and also changed the Redirect URL to my external HTTPS.

I restarted HA and in HACS I can now see this:
![image](https://user-images.githubusercontent.com/108185380/181134030-3a41a69d-5887-4f04-a5a7-1aef2a8ba198.png)
But I am not seeing the integration in Settings > Integrations so I cannot do any configuration steps. Nor do I see any entities or devices relating to Home Connect. That said, I am seeming this in the log:
2022-07-27 10:50:30 WARNING (SyncWorker_3) [homeassistant.loader] We found a custom integration home_connect_alt which has not been tested by Home Assistant. This component might cause stability problems, be sure to disable it if you experience issues with Home Assistant

I have not added the basic integration yet (it failed earlier as I didn't have the details in configuration.yaml). 

It feels like I am almost there. There are some references to _If you are modifying an existing Home Connect app_ but I am not doing this so I assume those notes should not apply.
My question is; should I add the basic Home Connect first? My assumption was No.
The machine is switched on and I have waited two hours.

Any thoughts? Needless to say, I am so grateful for the help you are giving me.

**Reply by llundberg57** [reply]
It seems like I have achieved nothing; if I go Settings > Devices & Services > Add Integration > Home Connect Alt, I get the same error.
{
  "error": "unauthorized_client",
  "error_description": "client or redirection URI not authorized",
  "state": "eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJmbG93X2lkIjoiMjg4M2M5M2M2NDIzNjNkZDNiYTQwZjZjYTI5MDBhZjAiLCJyZWRpcmVjdF91cmkiOiJodHRwczovL215LmhvbWUtYXNzaXN0YW50LmlvL3JlZGlyZWN0L29hdXRoIn0.QBjnHd8wl20Hnbw7QDxKcwAy0hOwaX06ivdGR4UWHwM"
}

This is AFTER I have added client_ID and client_secret to configuration.yaml with matching details from the Home Connect Developer website. Plus I have added the URL I use to access Home Assistant externally.
I believe I have followed all the instructions but I am at a dead end. All the other integrations done have gone easy and been straight forward but not this one.

**Comment by ekutner** [comment]
Post screenshots of your HC application configuration + the my.home-assistant.io page + your configuration. Once you have it working you can create an app with a different ID but as of now I really don't know how to help you, since it really shouldn't be this hard.

**Reply by llundberg57** [reply]
Hi again,
Here is a screenshot of the HC:

Here is a screenshot of what is in configuration.yaml:
![image](https://user-images.githubusercontent.com/108185380/181680262-a70a0488-b846-41f9-bb26-d445aa262ae5.png)

Screenshot of my.home-assistant.io page:
![image](https://user-images.githubusercontent.com/108185380/181680471-a097b19f-135a-49b1-8c79-1d6f311fb02f.png)

I have probably misunderstood something along the way.

Thanks again for your help. I do agree with you, I wish it wasn't this hard.

**Comment by ekutner** [comment]
As the instructions say pretty clearly, your redirect URI should be https://my.home-assistant.io/redirect/oauth
Just to confirm, when you navigate to homeassistant.local:8123 in your browser you get to your HA instance right?

Also, your name_template is not valid. I suggest you remove it from the configuration until you get this working and then try to read the explanation of how to use it more carefully.

**Reply by llundberg57** [reply]
Confirming that http://homeassistant.local:8123 does indeed get to my HA.
I have removed the name_template.
I have changed the redirect URl to https://my.home-assistant.io/redirect/oauth and progress, I can now see the integration.
![image](https://user-images.githubusercontent.com/108185380/181866464-ec3311cd-0f6f-441c-9eaa-e61947e160c2.png)

The only thing now is that the device I see seems generic (Home Connect Service) and the entities are also generic looking.
![image](https://user-images.githubusercontent.com/108185380/181866529-e607dc15-7d5d-44a3-b383-ec56631f558c.png)

I have switched on the dishwasher and will leave it for a while and see if it shows up. The dishwasher is the only Home Connect device I have so I am guessing that rate limit should not be the problem.
I turned wifi off on the phone and I could NOT see the device in the mobile Home Connect app which I think is correct?

Thanks again for your help, I know I have tested your patience so I apologise for that.

EDIT: for what its worth, I just saw this in the logs:

2022-07-30 13:44:53 WARNING (MainThread) [homeassistant.components.number] custom_components.home_connect_alt.number::OptionNumber is overriding deprecated methods on an instance of NumberEntity, this is not valid and will be unsupported from Home Assistant 2022.10. Please report it to the custom component author.
2022-07-30 13:44:53 WARNING (MainThread) [homeassistant.components.number] custom_components.home_connect_alt.number::SettingsNumber is overriding deprecated methods on an instance of NumberEntity, this is not valid and will be unsupported from Home Assistant 2022.10. Please report it to the custom component author

**Reply by Asdrubale1234** [reply]
i have this problem, "error": "unauthorized_client",
  "error_description": "Invalid client id"

my link http://homeassistant.local:8123/ and https://my.home-assistant.io/redirect/oauth

but I have this error, please help me

**Comment by ekutner** [comment, marked as answer]
you MUST see the appliance in the mobile app with WIFI TURNED OFF. If you do not see it then it is not properly connected to the HC service and you must go through the connection process again.

**Reply by llundberg57** [reply]
Thanks, 
That makes perfect sense. Just got it working. 
Thank you so much for your help and patience.

**Reply by PrestonMcAfee** [reply]
I am having trouble following this instruction.  I can see and control my Thermador (Bosch) dishwasher and oven in Home Connect, but when I turn wifi off, I get 'unable to connect locally'.  So I deleted the two appliances in the app, wiped the app data and cache, logged in with singlekey, added both appliances locally using the screens on the appliances. They still will not connect locally when I turn the wifi off.  I also tried connecting locally with cell data both on and off.  I hope someone can tell me what I am doing wrong.  I have not reset the appliances themselves.

**Reply by PrestonMcAfee** [reply]
For anyone experiencing the same problem, the HomeConnect app shows the ability to change the wifi SSID and password but in my experience this doesn't work -- the password is unchanged and fail to connect locally will persist.  I wound up resetting my oven and dishwasher at the devices and removing both from the HomeConnect app, and then reconnecting as if for the first time, which let me enter a new SSID and password.  I'm not sure when the oven and stove failed to connect locally or why and I hope it doesn't recur.  The oven is easy to reset but the dishwasher fairly time consuming.  

The HomeConnect app, in each device under settings, shows whether the device is connected locally (so you don't need to turn wifi off) to test, though that is a good test and that is where, in principle, you could change the wifi settings on the devices to connect locally, but as I said, this did not work in practice -- the wifi password showed as blank even after a new one was entered and the devices failed to connect locally.  

When I went to reload Home Connect Alt, it had already discovered both of my Home Connect devices.

If you are on the fence about switching from the Home Connect integration to Home Connect Alt, Home Connect Alt provides a lot more information, especially about the dishwasher.  Home Connect Alt is a much better integration and should be the default integration.
